async-query/src/main/java/org/opensearch/sql/spark/scheduler/job/OpenSearchRefreshIndexJob.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.spark.scheduler.job;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Singleton;
+import org.opensearch.jobscheduler.spi.JobExecutionContext;
+import org.opensearch.jobscheduler.spi.ScheduledJobParameter;
+import org.opensearch.jobscheduler.spi.ScheduledJobRunner;
+import org.opensearch.plugins.Plugin;
+import org.opensearch.sql.spark.scheduler.model.OpenSearchRefreshIndexJobRequest;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * The job runner class for scheduling refresh index query.
+ *
+ * <p>The job runner should be a singleton class if it uses OpenSearch client or other objects
+ * passed from OpenSearch. Because when registering the job runner to JobScheduler plugin,
+ * OpenSearch has not invoked plugins' createComponents() method. That is saying the plugin is not
+ * completely initialized, and the OpenSearch {@link org.opensearch.client.Client}, {@link
+ * ClusterService} and other objects are not available to plugin and this job runner.
+ *
+ * <p>So we have to move this job runner initialization to {@link Plugin} createComponents() method,
+ * and using singleton job runner to ensure we register a usable job runner instance to JobScheduler
+ * plugin.
+ */
+@Singleton
+public class OpenSearchRefreshIndexJob implements ScheduledJobRunner {
+
+  private static final Logger log = LogManager.getLogger(OpenSearchRefreshIndexJob.class);
+
+  public static OpenSearchRefreshIndexJob INSTANCE = new OpenSearchRefreshIndexJob();
+
+  public static OpenSearchRefreshIndexJob getJobRunnerInstance() {
+    return INSTANCE;
+  }
```

Comment:
With `@Singleton` annotation, Guice will use single instance when injecting to other classes. So we don't need to implement singleton pattern by ourselves.

Reply:
> With `@Singleton` annotation, Guice will use single instance when injecting to other classes. So we don't need to implement singleton pattern by ourselves.

getJobRunnerInstance is called during SQLPlugin instantiation, whereas `injector = modules.createInjector();` is then called in `createComponents`. Please let me know how to apply Guice in this case. Thanks!

Reply:
If injector is created after the instantiation, we cannot use `@Singleton` since that means we cannot rely on Guice to create the instance. But looks like the getJobRunnerInstance is called from `OpenSearchAsyncQueryScheduler`, which is instantiated in the same method. It should work even if we retrieve instance from the `injector` (retrieve instance of `OpenSearchAsyncQueryScheduler` from `injector`, and inject `OpenSearchRefreshIndexJob` into `OpenSearchAsyncQueryScheduler` by Guice).

Reply:
getJobRunnerInstance is a static function, and it's actually called https://github.com/opensearch-project/sql/pull/2834/files/2b69b66166dcc0b022404d7424717b6503265bef#diff-b9ca34b799db4aa7b3c7dbccba65de6f539a5d3d6accf8a1c75e4c6893b9a9b8R267 during SQLPlugin instantiation